compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2501,10 +2501,48 @@ object Types {
 
     /** A reference with the initial symbol in `symd` has an info that
      *  might depend on the given prefix.
+     *  Note: If M is an abstract type or non-final term member in trait or class C,
+     *  its info depends even on C.this if class C has a self type that refines
+     *  the info of M.
      */
     private def infoDependsOnPrefix(symd: SymDenotation, prefix: Type)(using Context): Boolean =
+
+      def refines(tp: Type, name: Name): Boolean = tp match
```

Comment:
Do we need a case for `TermRef` too?

Reply:
TermRefs are TypeProxies, so they are already handled.